libcxx/include/semaphore

```diff
@@ -95,19 +96,22 @@ public:
     _LIBCPP_ASSERT_ARGUMENT_WITHIN_DOMAIN(
         __update <= _LIBCPP_SEMAPHORE_MAX - __old, "update is greater than the expected value");
 
-    if (__old > 0) {
-      // Nothing to do
-    } else if (__update > 1)
+    if (__old == 0) {
```

Comment:
Is it possible to still keep the optimization like

```c++
if (__update > 1)
  __a_.notify_all();
else
  __a_.notify_one();
```

Reply:
So we can either do
```
if (__old == 0) {
  __a_.notify_all();
}
```
or as you suggested
```
if (__update > 1)
  __a_.notify_all();
else
  __a_.notify_one();
```

Both of them would solve the original lost wake up described in
https://bugs.llvm.org/show_bug.cgi?id=47013

It seems like that it is a trade off between
1. we don't need to notify anyone unless our counter reaches 0
2. we can just notify one thread if the caller passed `__update == 1`

Reply:
I agree, either approach should fix the problem. I wonder which one is better though. My gut feeling is that the first one is more in the "spirit of futex", i.e. that futex operations should be limited to the contended case. The downside of the second approach is that every `notify()` could lead to a system call, which might be bad for performance. I guess one would have to benchmark.

Sadly the ABI is locked in -- if one had an additional "waiters" field it would be possible to call `notify_one()`/`notify_all()` in an optimal way. A "lesser" ABI break could be to stuff both the number of waiters and the "permits" into the 64 bit of the semaphore (on 64 bit systems). Then, only `max()` and `least_max_value` would be affected. But who really needs 63/64 bit of semaphore range?

Reply:
> I agree, either approach should fix the problem. I wonder which one is better though. My gut feeling is that the first one is more in the "spirit of futex", i.e. that futex operations should be limited to the contended case. The downside of the second approach is that every `notify()` could lead to a system call, which might be bad for performance. I guess one would have to benchmark.
> 
> Sadly the ABI is locked in -- if one had an additional "waiters" field it would be possible to call `notify_one()`/`notify_all()` in an optimal way. A "lesser" ABI break could be to stuff both the number of waiters and the "permits" into the 64 bit of the semaphore (on 64 bit systems). Then, only `max()` and `least_max_value` would be affected. But who really needs 63/64 bit of semaphore range?

but we can't change the default template parameter number, can we?
`counting_semaphore<>` needs to stay `counting_semaphore<numeric_limits<ptrdiff_t>::max()>` , which means that all bits are used

Reply:
Thanks for the discussion. After discussing with @huixie90 , I understand better and I think

```c++
if (__old == 0) {
  __a_.notify_all();
}
```

is strictly better. I think avoiding an expensive syscall is probably most important. And our current implementation of `notify_one()` and `notify_all()` are actually the same right now, both do the equivalent of `notify_all()`. So the other alternative (`notify_all() / notify_one()`) doesn't save us anything.